subprojects/build-events/src/integTest/groovy/org/gradle/build/event/BuildEventsParallelIntegrationTest.groovy

```diff
@@ -81,4 +84,68 @@ class BuildEventsParallelIntegrationTest extends AbstractIntegrationSpec {
         then:
         noExceptionThrown()
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/24887")
+    @Issue("https://github.com/gradle/gradle/issues/27099")
+    def "build events listener service handles all events before it is closed"() {
+        buildFile("""
+            import ${BuildEventsListenerRegistry.name}
+            import ${OperationCompletionListener.name}
+            import ${FinishEvent.name}
+            import ${TaskFinishEvent.name}
+            import ${BuildServiceParameters.name}
+            import ${AtomicInteger.name}
+
+            abstract class BlockingListener implements OperationCompletionListener, BuildService<BuildServiceParameters.None>, Closeable {
+                private static final AtomicInteger instanceCount = new AtomicInteger()
+
+                BlockingListener() {
+                    System.out.println("listener \${instanceCount.incrementAndGet()} created")
+                }
+
+                @Override
+                void onFinish(FinishEvent event) {
+                    if (event instanceof TaskFinishEvent) {
+                        ${server.callFromBuildUsingExpression("'handle' + event.descriptor.taskPath")}
+                    } else {
+                        throw new IllegalArgumentException()
+                    }
+                }
+
+                @Override
+                void close() {
+                    ${server.callFromBuild("serviceClosed")}
+                }
+            }
+
+            // Do not inline listener, or it weakens the test.
+            // We try to create the shared services registry first to have it handle buildFinished earlier.
+            def listener = gradle.sharedServices.registerIfAbsent('listener', BlockingListener)
+            def registry = project.services.get(BuildEventsListenerRegistry)
+            registry.onTaskCompletion(listener)
+
+            tasks.register("a")
+            tasks.register("b") {
+                dependsOn("a")
+            }
+        """)
+
+        when:
+        def handleA = server.expectAndBlock("handle:a")
+        server.expect("handle:b")
+        server.expect("serviceClosed")
+
+        def build = executer.withTasks("b").startBuildProcessInDebugger(false).start()
```

Comment:
💅 Did you mean to keep the `startBuildProcessInDebugger(false)` call here?

Reply:
>Did you mean to keep the startBuildProcessInDebugger(false) call here?

Thank you for spotting!